components/raftstore/src/store/worker/raftlog_gc.rs

```diff
@@ -137,28 +109,39 @@ impl<EK: KvEngine, ER: RaftEngine> Runner<EK, ER> {
                     "end_index" => t.end_idx,
                 );
             }
-            groups.push(RaftLogGcTask {
-                raft_group_id: t.region_id,
-                from: t.start_idx,
-                to: t.end_idx,
-            });
-        }
-        let start = Instant::now();
-        match self.gc_raft_log(groups) {
-            Err(e) => {
+            if let Err(e) = self
+                .engines
+                .raft
+                .gc(t.region_id, t.start_idx, t.end_idx, &mut batch)
+            {
                 error!("failed to gc"; "err" => %e);
-                self.report_collected(0);
                 RAFT_LOG_GC_FAILED.inc();
             }
-            Ok(n) => {
-                debug!("gc log entries";  "entry_count" => n);
-                self.report_collected(n);
-                RAFT_LOG_GC_DELETED_KEYS_HISTOGRAM.observe(n as f64);
-            }
+        }
+        if let Err(e) = self.engines.raft.consume(&mut batch, false) {
```

Comment:
Maybe you can keep the most of original flush by:
```
fn gc_raft_log() {
   ...
   consume(batch)
   ...
}

fn flush() {
    ...
    for _ in _ {
        engine.gc(batch);
    }
    ...
}
```

So most code is unchanged and easier to understand.

Reply:
If it's the failpoint bothering you, the old code is simply wrong. It skips writing log batch but doesn't skip consuming callbacks. It is not the intended behavior as per https://github.com/tikv/tikv/blob/8864e9bac814554bbe4acaddd01282f5fdae3dae/tests/failpoints/cases/test_snap.rs#L701-L702

Reply:
Regardless of the behavior it is not what the test comment says, because tasks are lost forever. For RocksDB GC it means some entries will never be compacted again. So we will change the failpoint comment if that's what you prefer.

Reply:
The return action is to simulate the case that the node is stopped before completing GC. It's more appropriate to use "pause", but "pause" will stop the case from restarting node.

On the other hand, writing raft engine can fail. If that error happens, we just ignore it for now. I think they are two different things.